### PR TITLE
cost-model: add transaction cost conformance harness

### PR DESCRIPTION
**Pull request by cmoyes-jump**
#### Problem
Cross-client conformance testing for transaction cost accounting,

#### Summary of Changes
Adds the cost-model conformance harness to SVM, reusing the shared `feature_set_from_proto` converter. Exposes execute_cost (pure Rust) and `sol_compat_txn_cost_v1` (FFI).

#### Testing
Much like the other conformance harnesses, basic inline unit tests covering estimate happy path, and some edge cases missing txn, and oversized txn paths,

Part of issue #12378

### CODE REVIEW

File: svm/Cargo.toml [hunk 1]
```diff
@@ -78,6 +82,7 @@ solana-bpf-loader-program = { workspace = true, optional = true }
 solana-clock = { workspace = true }
 solana-compute-budget = { workspace = true, optional = true }
 solana-compute-budget-program = { workspace = true, optional = true }
+solana-cost-model = { workspace = true, optional = true }
```
Comment (buffalojoec) [inline svm/Cargo.toml]: This shouldn't be an SVM dependency. It's an Agave component.

File: svm/Cargo.toml [hunk 2]
```diff
@@ -94,13 +99,16 @@ solana-loader-v4-interface = { workspace = true }
...
 solana-nonce = { workspace = true }
 solana-nonce-account = { workspace = true, features = ["wincode"] }
+solana-packet = { workspace = true, optional = true }
 solana-precompile-error = { workspace = true, optional = true }
 solana-program-entrypoint = { workspace = true }
 solana-program-pack = { workspace = true }
 solana-program-runtime = { workspace = true }
 solana-pubkey = { workspace = true }
 solana-rent = { workspace = true }
+solana-runtime-transaction = { workspace = true, optional = true }
```
Comment (buffalojoec) [inline svm/Cargo.toml]: This one, too.

File: runtime/src/conformance/cost.rs [hunk 3]
```diff
@@ -0,0 +1,382 @@
...
+    },
+    solana_packet::PACKET_DATA_SIZE,
+    solana_pubkey::Pubkey,
+    solana_runtime_transaction::runtime_transaction::RuntimeTransaction,
+    solana_signature::Signature,
+    solana_transaction::{sanitized::MessageHash, versioned::VersionedTransaction},
+    std::ffi::c_int,
+};
+
+fn runtime_transaction_from_proto(
```
Comment (buffalojoec) [inline runtime/src/conformance/cost.rs:23]: This can probably be shared with #13211, right?

**Review by buffalojoec — CHANGES_REQUESTED** [review]
Generally the code looks good, but I think we should move this up to `solana-runtime`, not `solana-svm`. You're bringing in dependencies that live higher up in the validator's hierarchy - such as `solana-cost-model` and `solana-runtime-transaction`. Both of these are already dependencies in `solana-runtime`, and you can share some code with #13211 as well.

File: runtime/src/conformance/cost.rs (hunk 3 above)
Comment (cmoyes-jump) [reply runtime/src/conformance/cost.rs:23]: Added a TODO comment to consolidate, given #13211 hasn't been merged yet

File: runtime/src/conformance/cost.rs [hunk 4]
```diff
@@ -0,0 +1,388 @@
...
+        programs_execution_cost: cost.programs_execution_cost(),
+        loaded_accounts_data_size_cost: cost.loaded_accounts_data_size_cost(),
+        allocated_accounts_data_size: cost.allocated_accounts_data_size(),
+        total_cost: cost.sum(),
+    }
+}
+
+pub fn execute_cost(input: &CostContext) -> Option<CostResult> {
+    let proto_tx = input.tx.as_ref()?;
+    let runtime_tx = runtime_transaction_from_proto(proto_tx)?;
```
Comment (buffalojoec) [inline runtime/src/conformance/cost.rs]: Same question I asked @mjain-jump on #13211: do you want to use `.expect()` here and not use `Option` for the harness signature anymore?

File: runtime/src/conformance/cost.rs (hunk 3 above)
Comment (buffalojoec) [reply runtime/src/conformance/cost.rs:23]: Yours might be ready first, so we can maybe rebase #13211 onto this instead.

File: runtime/src/conformance/cost.rs [hunk 5]
```diff
@@ -0,0 +1,388 @@
...
+            actual_programs_execution_cost: 0,
+            actual_loaded_accounts_data_size_bytes: 0,
+        }
+    }
+
+    fn assert_has_cost(ctx: &CostContext) -> CostResult {
+        let result = execute_cost(ctx).expect("execute_cost returned None");
+        assert!(result.has_cost, "expected has_cost to be true");
+        result
+    }
```
Comment (buffalojoec) [inline runtime/src/conformance/cost.rs]: Kinda my point about the `Option`, eh?

**Review by buffalojoec — COMMENTED** [review]
LGTM! Just the question about the `Option`.

File: runtime/src/conformance/mod.rs [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+        MessageHeader, VersionedMessage,
+        compiled_instruction::CompiledInstruction,
+        legacy,
+        v0::{self, MessageAddressTableLookup},
+    },
+    solana_pubkey::Pubkey,
+};
+
+#[cfg(feature = "conformance")]
+pub(crate) fn build_versioned_message(value: &ProtoTransactionMessage) -> VersionedMessage {
```
Comment (buffalojoec) [inline runtime/src/conformance/mod.rs]: nit: can you rename this to `versioned_message_from_proto`?

File: runtime/src/conformance/cost.rs [hunk 7]
```diff
@@ -0,0 +1,322 @@
...
+    solana_signature::Signature,
+    solana_svm::conformance::feature_set::feature_set_from_proto,
+    solana_transaction::{sanitized::MessageHash, versioned::VersionedTransaction},
+};
+#[cfg(not(test))]
+use {prost::Message, std::ffi::c_int};
+
+fn runtime_transaction_from_proto(
+    proto_tx: &ProtoSanitizedTransaction,
+) -> Option<RuntimeTransaction<solana_transaction::sanitized::SanitizedTransaction>> {
```
Comment (buffalojoec) [inline runtime/src/conformance/cost.rs]: This function's return just gets unwrapped via `.expect`, so might as well lose the `Option` here, too.

Also nit: you can just import `SanitizedTransaction`.

File: runtime/src/conformance/cost.rs [hunk 8]
```diff
@@ -0,0 +1,322 @@
...
+    let proto_message = proto_tx.message.as_ref()?;
+    let message = build_versioned_message(proto_message);
+
+    let mut signatures: Vec<Signature> = proto_tx
+        .signatures
+        .iter()
+        .filter_map(|s| <[u8; 64]>::try_from(s.as_slice()).ok())
```
Comment (buffalojoec) [inline runtime/src/conformance/cost.rs]: I have been just using `expect` on these.
https://github.com/anza-xyz/agave/blob/4b1aa843b0f5ebdc484b829e567740ac8af5261a/svm/src/conformance/account_state.rs#L19

File: runtime/src/conformance/cost.rs [hunk 9]
```diff
@@ -0,0 +1,322 @@
+//! Transaction cost conformance harness.
+
+use {
+    super::build_versioned_message,
+    protosol::protos::{
+        CostContext, CostResult, SanitizedTransaction as ProtoSanitizedTransaction, TxnCostMode,
```
Comment (buffalojoec) [inline runtime/src/conformance/cost.rs]: nit: can you please prefix all of these with `Proto*` so it's clear which is which as you go through the module?

File: runtime/src/conformance/cost.rs [hunk 10]
```diff
@@ -0,0 +1,322 @@
...
+        MessageHash::Compute,
+        None,
+        SimpleAddressLoader::Disabled,
+        &std::collections::HashSet::new(),
+        true,
+    )
+    .ok()
+}
+
+fn result_from_transaction_cost<Tx>(cost: &TransactionCost<'_, Tx>) -> CostResult
```
Comment (buffalojoec) [inline runtime/src/conformance/cost.rs]: nit: `cost_result_to_proto` works, right?

**Review by buffalojoec — APPROVED** [review]
Awesome! Thanks for addressing all of my nits.
